### PR TITLE
ActionCable: allow to pass websocket.close args

### PR DESCRIPTION
**Pull request by jasl**
This PR allows you to set `code` and `reason` when closing a WebSocket connection.
It's very useful when hacking ActionCable.

In addition, I small refactor the file to avoid potential nil error, I can remove this change.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: actioncable/lib/action_cable/connection/web_socket.rb [hunk 1]
```diff
@@ -17,23 +17,23 @@ def possible?
...
       end
 
       def transmit(data)
-        websocket.transmit data
+        websocket&.transmit data
       end
 
-      def close
-        websocket.close
+      def close(code = nil, reason = nil)
```
Comment (palkan) [inline actioncable/lib/action_cable/connection/web_socket.rb:27]: This is an internal method, we do not expose it anywhere.

Thus, I'd suggest to hack around `Connection#close` in your codebase instead.

Comment (palkan) [reply actioncable/lib/action_cable/connection/web_socket.rb:27]: With this change, you still need to access internals, e.g., `self.websocket.close(code, reason)`, but you can already do this via `self.websocket.websocket.close(code, reason)`.

 > Could you give me some hints on how to hack Connection#close to customize the code and reason?
 
Smth like:
 
```ruby
def close(ws_code: nil, ws_reason: nil, **)
  return websocket.close(ws_code, ws_reason) if ws_code
  
  super(**)
end

# and then
connection.close(ws_code: 1001, ws_reason: "whatever")
```

Comment (jasl) [reply actioncable/lib/action_cable/connection/web_socket.rb:27]: I don't mean to change the `Connection#close` interface.

I can override the `Connection#close` method with my change.

```ruby
class MyConnection < ActionCable::Connection::Base
  def close
    code = 1001 # Determine which code should return
    websocket.close(code)
  end
end
```

Comment (jasl) [reply actioncable/lib/action_cable/connection/web_socket.rb:27]: In your code

```ruby
def close(ws_code: nil, ws_reason: nil, **)
  return websocket.close(ws_code, ws_reason) if ws_code
  
  super(**)
end
```
My patch is actually made this line `websocket.close(ws_code, ws_reason)` work.

The real `websocket` object allows passing `code` and `reason` on `close`,
it is wrapped by the `ActionCable::Connection::WebSocket` (the `websocket` instance variable in the `Connection` class), which doesn't expose the ability, I'm adding it.

**Comment by rafaelfranca** [comment]
This class isn't public API of the framework so it doesn't make much sense to add options so you can use in your application or library. This class should not be used in application or library because it isn't public.

Now, if this class should be private is another discussion. I'm not so sure. If we expect people to be interacting with that object it should not be. What reasons someone have to interact with this object?

**Comment by jasl** [comment]
> This class isn't public API of the framework so it doesn't make much sense to add options so you can use in your application or library. This class should not be used in application or library because it isn't public.
> 
> Now, if this class should be private is another discussion. I'm not so sure. If we expect people to be interacting with that object it should not be. What reasons someone have to interact with this object?

Returns a code, and a reason is WebSocket standard [Spec](https://www.rfc-editor.org/rfc/rfc6455.html#section-7.4.2) [MDN](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/close)

In my case, I'm extending ActionCable to add auto disconnect behavior if the client is idle long enough. The server side can give a clear reason to distinguish.

ActionCable now always returns code `1000`, which does not fully take advantage of WebSocket specification, so I think it is safe to expose the standard defined ability to the application level to not block advanced (or abnormal) usage, even though it's a private API.

**Comment by palkan** [comment]
> The server side can give a clear reason to distinguish would be great.

Then you should rely on Action Cable protocol here, i.e., send `{"type":"disconnect","reason":"idle_timeout"}` before closing. 

Using Action Cable implies following its conventions; if you want to support non-Action Cable-complient clients, you should use a custom connection class and implement your logic there.

> if this class should be private is another discussion

And this discussion happens in #50979 🙂 (tl;dr it should not; we should avoid making Connection/Channel classes dependent on the underlying transport mechanism)

**Comment by jasl** [comment]
> if you want to support non-Action Cable-complient clients, you should use a custom connection class and implement your logic there.

If I implement my own `Connection`, I need to access the `websocket` object to send custom messages, so it should not be private, and exposing `code` and `reason` seems reasonable, because it is standard.

In your PR, I replied I'm working on a PoC that implements the NoStr protocol based on the ActionCable connection.
https://github.com/jasl/dephy-relay/blob/main/app/channels/nostr_cable/connection.rb

**Comment by palkan** [comment]
>  exposing code and reason seems reasonable, because it is standard

For WebSockets, yes (and not widely used — that's why build custom protocols on top of them). I'd prefer to avoid coupling high-level Action Cable concepts with transport implementations. (But I'm not the one making decisions here; at least, yet, right, @rafaelfranca 😁?).

Anyway, the proposed change would allow to avoid the following hacking (kinda summary for those who might join the discussion and me thinking out load):

```ruby
websocket.send(:websocket).close(code, reason)
```

And instead have:

```ruby
websocket.close(code, reason)
```

The `websocket` object is currently available within a Connection class, it's `websocket.websocket` that is private. Make the latter one public definitely doesn't make any sense, it's wrapped for a reason (though I have no idea which one).

Since we name the wrapper object a "websocket", it sounds reasonable to support the `code` and `reason` arguments. Then, the only I would suggest is to change the proposed implementation to the following:

```ruby
def close(...)
  websocket.close(...)
end
```

**Comment by rafaelfranca** [comment]
> For WebSockets, yes (and not widely used — that's why build custom protocols on top of them). I'd prefer to avoid coupling high-level Action Cable concepts with transport implementations. (But I'm not the one making decisions here; at least, yet, right, @rafaelfranca 😁?).

If decision are well based like this one, I'm glad that you are trying to make it.

It seems we should expose that websocket object. I agree we should be using the `(...)` operator as well, so can you change that?

**Comment by jasl** [comment]
> > For WebSockets, yes (and not widely used — that's why build custom protocols on top of them). I'd prefer to avoid coupling high-level Action Cable concepts with transport implementations. (But I'm not the one making decisions here; at least, yet, right, @rafaelfranca 😁?).
> 
> If decision are well based like this one, I'm glad that you are trying to make it.
> 
> It seems we should expose that websocket object. I agree we should be using the `(...)` operator as well, so can you change that?

Changed, and I also refactored `transmit(data)` with the same style. If you don't like it, I can revert

**Review by palkan — COMMENTED** [review]
Thanks @jasl!

LGTM 👍

**Pull request merged**